torch/nn/functional.py

```diff
@@ -5182,6 +5196,29 @@ def multi_head_attention_forward(
             check_other=False,
         )
 
+        if key_padding_mask is not None:
+            # this is where we have atrociously painted ourselves into a corner
+            # when a key_padding mask is specified, it is merged with attn_mask
+            # but here we don't have the attn_mask, only is causal.
+            # Options:
+            # 1 - instantiate causal attn_mask and live happily ever after (perf will suck)
+            # 2 - allow specifying key_padding_mask to SDPA separately
+            # 3 - convert input to nested tensor to pake in kpm
+
+            # Option (1) for now.  Not sure I have the right dimensions
+            # FIXME:@drisspg
+            # if is_causal and attn_mask is None:
+            #     L = query.size(-2)
+            #     S = key.size(-2)
+            #     attn_mask = torch.triu(torch.full((L, S), float('-inf'), device=query.device, dtype=query.dtype), diagonal=1)
+
+            assert not (is_causal and attn_mask is None), "FIXME: instantiate causal attn_mask"
```

Comment:
If this is an error that the user will see, we should have a nicer error message.

Reply:
The user should not.  This was a debug check.  We have two options - we can either require attn_mask to always be supplied, or instantiate it in cases where we need.  

I settled on conceptually using is_)causal as a flag about the nature of attn_mask, and always providing attn_mask is conceptually simpler.  (Basically, from a user perspective, you can specify is_causal to get better perf, or just keep using the legacy API.)